File: go-backend-interview-prep/05_errors_and_testing/10_benchmarks_intro/main.go
package main

import "fmt"

// Fib computes the n-th Fibonacci number recursively.
func Fib(n int) int {
	if n <= 1 {
		return n
	}
	return Fib(n-1) + Fib(n-2)
}

// FibIter computes the n-th Fibonacci number iteratively.
func FibIter(n int) int {
	if n <= 1 {
		return n
	}
	a, b := 0, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b
}

// StringConcat builds a string by concatenation (slow).
func StringConcat(n int) string {
	s := ""
	for i := 0; i < n; i++ {
		s += "a"
	}
	return s
}

func main() {
	fmt.Println("Fib(10) =", Fib(10))
	fmt.Println("FibIter(10) =", FibIter(10))
	fmt.Println("Fib(20) =", Fib(20))
	fmt.Println("FibIter(20) =", FibIter(20))
	fmt.Println("len(StringConcat(100)) =", len(StringConcat(100)))
}

File: go-backend-interview-prep/05_errors_and_testing/10_benchmarks_intro/main.py
"""Benchmarks intro -- Python equivalent of the Go example."""


def fib(n: int) -> int:
    """Recursive Fibonacci."""
    if n <= 1:
        return n
    return fib(n - 1) + fib(n - 2)


def fib_iter(n: int) -> int:
    """Iterative Fibonacci."""
    if n <= 1:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b


def string_concat(n: int) -> str:
    """Build a string by repeated concatenation (slow)."""
    s = ""
    for _ in range(n):
        s += "a"
    return s


def main() -> None:
    print("fib(10) =", fib(10))
    print("fib_iter(10) =", fib_iter(10))
    print("fib(20) =", fib(20))
    print("fib_iter(20) =", fib_iter(20))
    print("len(string_concat(100)) =", len(string_concat(100)))


if __name__ == "__main__":
    main()
